server/coordinator/watch/watch.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+package coordinator
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/CeresDB/ceresmeta/pkg/log"
+	"github.com/CeresDB/ceresmeta/server/storage"
+	"github.com/pkg/errors"
+	"go.etcd.io/etcd/api/v3/mvccpb"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.uber.org/zap"
+)
+
+const (
+	shardPath = "shards"
+)
+
+type ShardRegisterEvent struct {
+	ShardID       storage.ShardID
+	NewLeaderNode string
+}
+
+type ShardExpireEvent struct {
+	ShardID       storage.ShardID
+	OldLeaderNode string
+}
+
+type ShardEventCallback interface {
+	OnShardRegistered(event ShardRegisterEvent) error
+	OnShardExpired(event ShardExpireEvent) error
+}
+
+// ShardWatch used to watch the distributed lock of shard, and provide the corresponding callback function.
+type ShardWatch struct {
+	rootPath       string
+	etcdClient     *clientv3.Client
+	eventCallbacks []ShardEventCallback
+
+	lock      sync.RWMutex
+	isRunning bool
+	quit      chan struct{}
+}
+
+func NewWatch(rootPath string, client *clientv3.Client) *ShardWatch {
+	return &ShardWatch{
+		rootPath:       rootPath,
+		etcdClient:     client,
+		eventCallbacks: []ShardEventCallback{},
+		quit:           make(chan struct{}),
+	}
+}
+
+func (w *ShardWatch) Start(ctx context.Context) error {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+
+	if w.isRunning {
+		return nil
+	}
+
+	shardsKeyPath := strings.Join([]string{w.rootPath, shardPath}, "/")
+
+	if err := w.startWatch(ctx, shardsKeyPath); err != nil {
+		return errors.WithMessage(err, "etcd register watch failed")
+	}
+
+	w.isRunning = true
+	return nil
+}
+
+func (w *ShardWatch) Stop(ctx context.Context) error {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+
+	w.isRunning = false
+	w.quit <- struct{}{}
+	return nil
+}
+
+func (w *ShardWatch) RegisteringEventCallback(eventCallback ShardEventCallback) {
+	w.eventCallbacks = append(w.eventCallbacks, eventCallback)
+}
+
+func (w *ShardWatch) startWatch(ctx context.Context, path string) error {
+	log.Info("register shard watch", zap.String("watchPath", path))
+	go func() {
+		for {
+			select {
+			case <-w.quit:
+				return
+			default:
+				respChan := w.etcdClient.Watch(ctx, path, clientv3.WithPrefix(), clientv3.WithPrevKV())
+				for resp := range respChan {
+					for _, event := range resp.Events {
+						if err := w.processEvent(event); err != nil {
+							log.Error("process event", zap.Error(err))
+						}
+					}
+				}
+			}
```

Comment:
Is it able to quit when `quit` message is received during the watching procedure.